crates/biome_deserialize/src/lib.rs

```diff
@@ -1,31 +1,346 @@
+    /// Any diagnostics emitted during deserialization are appended to `diagnostics`.
+    fn deserialize(
+        value: impl DeserializableValue,
+        diagnostics: &mut Vec<DeserializationDiagnostic>,
+    ) -> Option<Self>;
+}
+
+/// Implemented by data structure that can be deserialized.
+///
+/// This trait should only be implemented when adding the support for a new data format.
+/// See [biome_deserialize::json] for an example of implementation.
+pub trait DeserializableValue: Sized {
+    /// Range in the source content of this value
+    fn range(&self) -> TextRange;
+
+    /// Returns the deserialized form of this value using `visitor`.
+    /// Any diagnostics emitted during deserialization are appended to `diagnostics`.
+    fn deserialize<V: DeserializationVisitor>(
+        self,
+        visitor: V,
+        diagnostics: &mut Vec<DeserializationDiagnostic>,
+    ) -> Option<V::Output>;
+}
+
+/// This trait represents a visitor that walks through a [DeserializableValue].
+///
+/// We assume that a deserializable value has one of the following type:
+/// - null / none
+/// - boolean
+/// - number (integer, floats)
+/// - string
+/// - array
+/// - map (key-value pairs with value's type that can depend on its key)
+///
+/// Every type is associated to a `visit_` method.
+/// [DeserializableValue::deserialize] calls the `viist_` method that matches the type of the value.
+///
+/// Most of the time you should implement [Deserializable] and rely on existing deserializable types.
+/// You should use a visitor only when you deserialize a map or a union of several types.
+///
+/// ### Examples
+///
+/// ```
+/// use biome_deserialize::{DeserializationDiagnostic,  Deserializable, DeserializableValue, DeserializationVisitor, ExpectedType};
+/// use biome_deserialize::json::deserialize_from_json_str;
+/// use biome_rowan::{TextRange, TokenText};
+///
+/// struct Person {
+///     name: String
+/// }
+///
+/// impl Deserializable for Person {
+///     fn deserialize(
+///         value: impl DeserializableValue,
+///         diagnostics: &mut Vec<DeserializationDiagnostic>,
+///     ) -> Option<Self> {
+///         value.deserialize(PersonVisitor, diagnostics)
+///     }
+/// }
+///
+/// struct PersonVisitor;
+/// impl DeserializationVisitor for PersonVisitor {
+///     type Output = Person;
+///     const EXPECTED_TYPE: ExpectedType = ExpectedType::MAP;
+///
+///     fn visit_map(
+///         self,
+///         members: impl Iterator<Item = (impl DeserializableValue, impl DeserializableValue)>,
+///         range: TextRange,
+///         diagnostics: &mut Vec<DeserializationDiagnostic>,
+///     ) -> Option<Self::Output> {
+///         const ALLOWED_KEYS: &[&str] = &["name"];
+///         let mut name = None;
+///         for (key, value) in members {
+///             let key_range = key.range();
+///             let Some(key) = TokenText::deserialize(key, diagnostics) else {
```

Comment:
For simplicity, can we use `String::deserialize` instead? It's an example, and using a `TokenText` without an explanation can cause confusion.